examples/adwords_axis/src/main/java/adwords/axis/v201710/campaignmanagement/AddCompleteCampaignsUsingBatchJob.java

```diff
@@ -0,0 +1,452 @@
+      List<AdGroupOperation> adGroupOperations) {
+    List<AdGroupAdOperation> operations = new ArrayList<>();
+    for (AdGroupOperation adGroupOperation : adGroupOperations) {
+      long adGroupId = adGroupOperation.getOperand().getId();
+      AdGroupAd adGroupAd = new AdGroupAd();
+      adGroupAd.setAdGroupId(adGroupId);
+
+      ExpandedTextAd textAd = new ExpandedTextAd();
+      textAd.setHeadlinePart1("Luxury Cruise to Mars");
+      textAd.setHeadlinePart2("Visit the Red Planet in style.");
+      textAd.setDescription("Low-gravity fun for everyone!");
+      textAd.setFinalUrls(new String[] {"http://www.example.com/1"});
+
+      adGroupAd.setAd(textAd);
+
+      AdGroupAdOperation operation = new AdGroupAdOperation();
+      operation.setOperator(Operator.ADD);
+      operation.setOperand(adGroupAd);
+
+      operations.add(operation);
+    }
+    return operations;
+  }
+
+  private static List<AdGroupCriterionOperation> buildAdGroupCriterionOperations(
+      List<AdGroupOperation> adGroupOperations) {
+    List<AdGroupCriterionOperation> adGroupCriteriaOperations = new ArrayList<>();
+
+    // Create AdGroupCriterionOperations to add keywords.
+    for (AdGroupOperation adGroupOperation : adGroupOperations) {
+      long newAdGroupId = adGroupOperation.getOperand().getId();
+      for (int i = 0; i < NUMBER_OF_KEYWORDS_TO_ADD; i++) {
+        // Create Keyword.
+        String text = String.format("mars%d", i);
+
+        // Make 50% of keywords invalid to demonstrate error handling.
+        if (i % 2 == 0) {
+          text = text + "!!!";
+        }
+        Keyword keyword = new Keyword();
+        keyword.setText(text);
+        keyword.setMatchType(KeywordMatchType.BROAD);
+
+        // Create BiddableAdGroupCriterion.
+        BiddableAdGroupCriterion biddableAdGroupCriterion = new BiddableAdGroupCriterion();
+        biddableAdGroupCriterion.setAdGroupId(newAdGroupId);
+        biddableAdGroupCriterion.setCriterion(keyword);
+
+        // Create AdGroupCriterionOperation.
+        AdGroupCriterionOperation operation = new AdGroupCriterionOperation();
+        operation.setOperand(biddableAdGroupCriterion);
+        operation.setOperator(Operator.ADD);
+
+        // Add to list.
+        adGroupCriteriaOperations.add(operation);
+      }
+    }
+    return adGroupCriteriaOperations;
+  }
+
+  private static List<AdGroupOperation> buildAdGroupOperations(Iterator<Long> tempIdGenerator,
+      String namePrefix, Iterable<CampaignOperation> campaignOperations) {
+    List<AdGroupOperation> operations = new ArrayList<>();
+    for (CampaignOperation campaignOperation : campaignOperations) {
+      for (int i = 0; i < NUMBER_OF_ADGROUPS_TO_ADD; i++) {
+        AdGroup adGroup = new AdGroup();
+        adGroup.setCampaignId(campaignOperation.getOperand().getId());
+        adGroup.setId(tempIdGenerator.next());
+        adGroup.setName(String.format("Batch Ad Group %s.%s", namePrefix, i));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: i (declared at line 352). This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 356. Consider formatting this data with &#39;%d&#39; instead.